web/pandas/pdeps/00xx-string-dtype.md

```diff
@@ -0,0 +1,253 @@
+
+Since its introduction, the `StringDtype` has always been opt-in, and has used
+the experimental `pd.NA` sentinel for missing values (which was also [introduced
+in pandas 1.0](https://pandas.pydata.org/docs/whatsnew/v1.0.0.html#experimental-na-scalar-to-denote-missing-values)).
+However, up to this date, pandas has not yet taken the step to use `pd.NA` by
+default, and thus the `StringDtype` deviates in missing value behaviour compared
+to the default data types.
+
+In 2023, [PDEP-10](https://pandas.pydata.org/pdeps/0010-required-pyarrow-dependency.html)
+proposed to start using a PyArrow-backed string dtype by default in pandas 3.0
+(i.e. infer this type for string data instead of object dtype). To ensure we
+could use the variant of `StringDtype` backed by PyArrow instead of Python
+objects (for better performance), it proposed to make `pyarrow` a new required
+runtime dependency of pandas.
+
+In the meantime, NumPy has also been working on a native variable-width string
+data type, which will be available [starting with NumPy
+2.0](https://numpy.org/devdocs/release/2.0.0-notes.html#stringdtype-has-been-added-to-numpy).
+This can provide a potential alternative to PyArrow for implementing a string
+data type in pandas that is not backed by Python objects.
+
+After acceptance of PDEP-10, two aspects of the proposal have been under
+reconsideration:
+
+- Based on user feedback (mostly around installation complexity and size), it
+  has been considered to relax the new `pyarrow` requirement to not be a _hard_
+  runtime dependency. In addition, NumPy 2.0 could in the future potentially
+  reduce the need to make PyArrow a required dependency specifically for a
+  dedicated pandas string dtype.
+- The PDEP did not consider the usage of the experimental `pd.NA` as a
+  consequence of adopting one of the existing implementations of the
+  `StringDtype`.
+
+For the second aspect, another variant of the `StringDtype` was
+[introduced in pandas 2.1](https://pandas.pydata.org/docs/whatsnew/v2.1.0.html#whatsnew-210-enhancements-infer-strings)
+that is still backed by PyArrow but follows the default missing values semantics
+pandas uses for all other default data types (and using `NaN` as the missing
+value sentinel) ([GH-54792](https://github.com/pandas-dev/pandas/issues/54792)).
+At the time, the `storage` option for this new variant was called
+`"pyarrow_numpy"` to disambiguate from the existing `"pyarrow"` option using `pd.NA`.
+
+This last dtype variant is what you currently (pandas 2.2) get for string data
+when enabling the ``future.infer_string`` option (to enable the behaviour which
+is intended to become the default in pandas 3.0).
+
+## Proposal
+
+To be able to move forward with a string data type in pandas 3.0, this PDEP proposes:
+
+1. For pandas 3.0, we enable a "string" dtype by default, which will use PyArrow
+   if installed, and otherwise falls back to an in-house functionally-equivalent
+   (but slower) version.
```

Comment:
Isn't the "in-house functionally-equivalent (but slower) version" the current implementation based on numpy 1.x in version 2.2, but we now make the dtype `string` instead of `object` ?

Reply:
In the section below that expands on this object-type based implementation ("Object-dtype "fallback" implementation"), there is a bit longer explanation and I also link to the open PR implementing this: https://github.com/pandas-dev/pandas/pull/58451

(and fwiw, this is not specific to numpy 1.x, it will also work on numpy 2.x, it just uses `object` dtype)

Reply:
I have to admit that I am pretty confused by these various implementations and the corresponding semantics used to describe them.

I'd suggest having a summary of what exists today in pandas 2.2 (strings with object dtype, `np.nan`), the current `StringDtype` with `pd.NA`, the "experimental" `pyarrow` based implementations (with both `pd.NA` and `np.nan` being available), and anything else, and what is proposed would be available due to this PDEP, and how it might change in the future due to however we decide to handle missing values in the future, as well as `nanoarrow` and `numpy 2.0` strings.

Reply:
Current situation (pandas 2.2) is:

* Current default: `object` dtype with np.nan or None
* Experimental opt-in string dtypes using `pd.NA`: `StringDtype()` with storage being `"python"` (default, object dtype under the hood) or `"pyarrow"`
* Future string dtypes using `NaN` (behind `pd.options.future.infer_string = True`): `StringDtype()` with storage being `"pyarrow_numpy"`
  * This is the dtype that is essentially being proposed in this PDEP, but already exists since pandas 2.1 ([#54792](https://github.com/pandas-dev/pandas/issues/54792))

And then this PDEP also describes to add an extra option for the third bullet point, i.e. having a `StringDtype()` using NaN but backed by object-dtype array instead of pyarrow, which I dubbed (for now) `"python_numpy"` in the PR adding this (https://github.com/pandas-dev/pandas/pull/58451), but that name is still being discussed.

As a starter (still thinking about how to make this clearer in the PDEP), does the above clarify it for you?

Note that the above listing leaves out `pd.ArrowDtype(<some pyarrow string type>)`, because that is not really relevant for the discussion right now and I am not sure mentioning it will help, but that is yet another way to store strings.

Reply:
> As a starter (still thinking about how to make this clearer in the PDEP), does the above clarify it for you?
> 
Yes, although I'd suggest for the PDEP a table that indicates the constructor version of the dtype (e.g., `dtype=object`, or `dtype=pd.StringDtype()`), and the string representation of the dtype.

> Note that the above listing leaves out `pd.ArrowDtype(<some pyarrow string type>)`, because that is not really relevant for the discussion right now and I am not sure mentioning it will help, but that is yet another way to store strings.

Is that equivalent to having a pure `"pyarrow"` backed string that uses `pd.NA` as the null semantics?  If not, where does that fit it?  Or is it not available?

Reply:
> Is that equivalent to having a pure `"pyarrow"` backed string that uses `pd.NA` as the null semantics? If not, where does that fit it? Or is it not available?

`StringDtype(storage="pyarrow")` is also a "pure pyarrow backed string that uses NA", to be clear. But so 
`StringDtype(storage="pyarrow")`and `ArrowDtype(pa.(large_)string())` are essentially equivalent, except for using a different dtype and array class.